game-core/src/main/java/games/strategy/engine/framework/message/PlayerListing.java

```diff
@@ -39,15 +43,19 @@
    * Creates a new instance of PlayerListing.
    */
   public PlayerListing(final Map<String, String> playerToNodeListing, final Map<String, Boolean> playersEnabledListing,
-      final Map<String, String> localPlayerTypes, final Version gameVersion, final String gameName,
+      final Map<String, PlayerType> localPlayerTypes, final Version gameVersion, final String gameName,
       final String gameRound, final Collection<String> playersAllowedToBeDisabled,
       final Map<String, Collection<String>> playerNamesAndAlliancesInTurnOrder) {
     m_playerToNodeListing =
         playerToNodeListing == null ? new HashMap<>() : new HashMap<>(playerToNodeListing);
     m_playersEnabledListing = playersEnabledListing == null ? new HashMap<>()
         : new HashMap<>(playersEnabledListing);
-    m_localPlayerTypes =
-        localPlayerTypes == null ? new HashMap<>() : new HashMap<>(localPlayerTypes);
+    m_localPlayerTypes = (localPlayerTypes == null)
+        ? new HashMap<>()
+        : localPlayerTypes.entrySet()
+            .stream()
+            // convert Map<String,PlayerType> -> Map<String,String>
+            .collect(Collectors.toMap(Entry::getKey, e -> e.getValue().getLabel()));
```

Comment:
Note that this could cause problems if the map contains null values

Reply:
Good concern. Looking at it, it's hard to spot any places where we would add a null. The existing code is defensive about null references, but if we get a map it's more reasonable to expect it has values. Considering as part of the model we need a player type for each player selection, it is reasonable for us to expect we always have this.

A second check, which convinced me we are okay, is the existing code would have NPE'd if there were any null map value entries. Note the code below from [`Triplea.java`](https://github.com/triplea-game/triplea/pull/3592/files#diff-2bd95d259055fc3b03aaf2f1522db151L64):

```
 @Override
-  public Set<IGamePlayer> createPlayers(final Map<String, String> playerNames) {
-    final Set<IGamePlayer> players = new HashSet<>();
-    for (final String name : playerNames.keySet()) {
-      final String type = playerNames.get(name);
-      if (type.equals(WEAK_COMPUTER_PLAYER_TYPE)) {
```

************

I think there is reasonable room for improvement, I added: 87631f9cc374fa35e472f1d64a93be80912ae0ab
- converts null syntax to use optionals
- does a precondition check asserting no map values have null checks. We'll be pretty sure i think to find in testing if our assumption fails.